IPIP-499: UnixFS CID Profiles (#499)
ipfs/specs: src/ipips/ipip-0499.md

---
# IPIP number should  match its pull request number. After you open a PR,
# please update title and update the filename to `ipip0000`.
title: "IPIP-0499: CID Profiles"
date: 2025-04-03
ipip: proposal
editors:
  - name: Michelle Lee
relatedIssues:
  - n/a
order: 0000
tags: ['ipips']
---

## Summary

<!--One paragraph explanation of the IPIP.-->
This proposal introduces profiles for IPFS CIDs. Profiles explicitly define CID version, hash algorithm, chunk size, DAG width, layout, and other parameters. 

## Motivation

Currently, CIDs can be generated with a variety of settings and optimizations for chunking, DAG width, and more. This means the same file can yield multiple, different CIDs depending on which tools and settings are used, and it is not possible to reliably reproduce or verify the CID. Profiles offer With profiles, following the same profile will produce identical CIDs for identical content, whic makes verification regardless of implementation.

## Detailed design

We introduce a profile naming system, 

Each profile must specify the following characteristics:

1. CID version (CIDv0 or CIDv1)
2. Hash algorithm
3. Chunk size
4. DAG width

> 5. DAG layout

Reviewer: I'd add the following parameters:
- HAMT threshold (when a block will be converted into a HAMT)
- HAMT fanout, aka HAMT dag width (all known examples use 256)
- Raw blocks (vs dag-pb file blocks)
- Are empty directories added by default? 
-

> 1. CID version (CIDv0 or CIDv1)
> 2. Hash algorithm
> 3. Chunk size
> 4. DAG width
> 5. DAG layout

Reviewer 2: ```suggestion
1. CID version (currently only CIDv0 or CIDv1)
2. Hash algorithm
3. UnixFS Chunk size (explicitly set, not contextual/reactive to input)
4. UnixFS directory DAG width
5. UnixFS directory DAG layout
6. HAMT directory DAG threshold
7. HAMT directory DAG width
8. Leaf Envelope (historically dag-pb, raw now)
9. Allow empty directories
```

> 6. Required

Reviewer 2: I think there's a word missing here 😅

Additional profiles can be added at a future date. Profile names may be chosen from the names of any botanical tree with compound leaves.

> |             | Helia default | Kubo default                | Storacha default | "test-cid-v1" profile | DASL          |

Reviewer 3: This is only one of Kubo presets – basically "CIDv0" as CIDv1 + raw leaves. 
https://github.com/ipfs/kubo/pull/10774 renames `test-cid-v1` to `legacy-cid-v1` to make it more clear (I'll merge this comment once it lands)

```suggestion
|             | Helia default | Kubo default                | Storacha default | Kubo `legacy-cid-v1` profile | DASL          |
```

|-------------|---------------|-----------------------------|------------------|-----------------------|---------------|
| CID version | CIDv1         | CIDv1                       | CIDv1            | CIDv1                 | CIDv1         |
| Hash Algo   | sha-256       | sha-256                     | sha-256          | sha-256               | sha-256       |
| Chunk size  | 1MiB          | 256KiB                      | 1MiB             | 1MiB                  | not specified |
| DAG width   | 1024          | 174 (but it's complicated*) | 1024             | 174                   | not specified |
| DAG layout  | balanced      | balanced                    | balanced         | balanced              | not specified |

> 

Reviewer 2: This is a good informational thing to show defaults are different, but what probably matters most to both end-users and implementers/backlog-managers is whether defaults can be overridden per input (e.g. by CLI flag) or per environment (e.g. by configuration/env var), or both.  It's totally fine for conformance to either profile to require extra work and not be the default, but I'm worried that even the affordance to override defaults for many of these settings lingers indefinitely in backlogs, or get more onerous to implement as time goes by!

> 

Reviewer 3: Good point, HAMT size threshold should be added.

fysa https://github.com/ipfs/kubo/pull/10774 aims to expose  `Import.UnixFSHAMTDirectorySizeThreshold` and I think we should agree how `cid-v1-2025` should look like: if we want to keep HAMT chunks at 256KiB, or if we want to align them with file chunk size which is 1MiB.

| Max links `Directory` node      | ?             | 0                                 | ?                | 0                  | 0                         | ?             |
| Max fanout `HAMTDirectory` node | 256 blocks    | 256 blocks                        | 256 blocks       | 256 blocks         | **1024**                  | not specified |
| `HAMTDirectory` threshold       | 256KiB (est)  | 256KiB (est:links[name+cid])      | 1000 **links**   | 256KiB             | **1MiB**                  | not specified |
| DAG layout                      | balanced      | balanced                          | balanced         | balanced           | balanced                  | not specified |
| Leaves                          | raw           | raw                               | raw              | raw                | raw                       | not specified |
| Empty directories               | allowed       | allowed                           | disallowed       | allowed            | allowed                   | not specified |

See related discussion at https://discuss.ipfs.tech/t/should-we-profile-cids/18507/

### User benefit

> Reliable, deterministic CIDs allow independent verification of content across tools and ipmlementations.

Reviewer 4: ```suggestion
Reliable, deterministic CIDs allow independent verification of content across tools and implementations.
```

### Compatibility

Implementations will need to (1) make CID generation settings configurable and (2) support user setting of profiles.

Kubo 0.35 will have [`Import.*` configuration](https://github.com/ipfs/kubo/blob/master/docs/config.md#import) option to control DAG width.

### Security

TODO

### Alternatives

Another approach could be to name profiles based on the key UnixFS/CID parameters, e.g. v1-sha256-balanced-1mib-1024w-raw. This is longer and more convoluted.

#### Empty directories

> Decision if empty directories should be included is left out of scope.

Reviewer: @lidel I think we should include this, because the goal of this spec is to make UnixFS encoding deterministic when using the profile, and this stands in the way of this, potentially rendering this whole effort futile if not included. 

I think the profile should dictate whether empty directories are included. This should also be configurable such that you can adjust to your own needs.

By introducing profiles, we can benefit from both the optionality offered by UnixFS, where users are free to chose their own parameters, and determinism through profiles.

UnixFS CIDs can be generated with a variety of settings and optimizations for chunking, DAG width, and more. This means the same file tree can yield multiple, different CIDs depending on which tools and settings are used, and it is not possible to reliably reproduce or verify the CID. Profiles offer With profiles, following the same profile will produce identical CIDs for identical content, whic makes verification regardless of implementation.

## Detailed design

We introduce a set of named profiles that define a set of parameters for generating UnixFS CIDs. These profiles can be used by implementations to ensure that the same content will yield the same CID across different tools and implementations.

### UnixFS parameters

The profiles define a set of parameters that affect the resulting CID. These parameters are based on the UnixFS specification and are used to generate the CID for a given file tree. The parameters include:

> 1. CID version (currently only CIDv0 or CIDv1)

Reviewer: Good point. I think we should dictate the multibase too. The final string representation of the CID matters here.

> 1. CID version (currently only CIDv0 or CIDv1)

Reviewer 5: dagpb encodes the whole cid bytes, so version matters

> 1. Hash function

Reviewer: That's right. I think this IPIP should dictate the hash function for all nodes in the DAG to meet the goal of CID determinism. That's the point of the profile. This is a subset of UnixFS after all.

1. UnixFS chunk size

> 1. UnixFS DAG layout (e.g. balanced, trickle)

Reviewer: > We probably need to fix the unixfs spec as it doesn't really say how the layout works. 

Yeah, maybe a point to add to https://github.com/ipfs/specs/pull/331

> For a start, they DAG layout is not part of the unixfs "format", but rather unixfs protobufs are embedded in merkledag-pb protobufs, that could be cbors or jsons. So either we set that in stone (unixfs dag == protobuf-pb) or we need to be explicit in profiles.

I think I understand. Do you mean that the DAG layout is a function of the `PBNode.Links` which are on the dag-pb layer, i.e. outer envelope of UnixFS?

> Then there is the part about the subtleties of the layout. i.e. the trickle parameters, the balancing of the balanced dag etc, which are also badly specified. It may be that now you can get away with UnixFS DAG width for configuring existing DAG layouts, but I would imagine layouts with more options than the width.

Right. We should refine this section a bit more and keep it open ended. But how do we avoid a combinatorial explosion to account for all possibilities? 

This IPIP introduces a single named profile, `unixfs-2025` and the minimum necessary parameters to define it. 

My question to you is if you actually think it's necessary to already introduce profiles for other DAG layouts?

1. UnixFS DAG width (max number of links per `File` node)

> 1. `HAMTDirectory` fanout (must be a power of 2)

Reviewer: Bit-width is a bit more intuitive imo, because it ties more deeply to how the HAMT is implemented.

> 1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on an estimate of the block size by counting the size of PNNode.Links

Reviewer 2: wait is kubo dynamically assigning HAMT Directory threshold, currently? i was assuming this was a static number!

> 1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on an estimate of the block size by counting the size of PNNode.Links

Reviewer 6: If you want to be exact you have to take into account any non-zero value fields in the serialized root UnixFS metadata since these affect the block size.

It's quite possible that Kubo will produce a HAMT block that's too big with a certain combination of directory entry names if someone has also changed the encoded directory's default mtime or whatever, probably because the "should-I-shard" feature pre-dates Kubo's ability to add UnixFSv1.5 metadata to things.

Really there's no need to estimate anything - it's trivial to count the actual bytes that a block will take up and then shard if necessary.

1. Leaf Envelope: either `dag-pb` or `raw`
1. Whether empty directories are included in the DAG

>   - Some implementations apply filtering before merkleizing filesystem entries in the DAG.

Reviewer: This is motivated by Git's default behaviour which ignores empty directories.

>   - Some implementations apply filtering before merkleizing filesystem entries in the DAG.

Reviewer 7: Do we have existing implementations that support filtering differently on all of these? Because unless we do, I would really rather not specify all possible variants. And I agree with @bumblefudge: let's have two behaviours if possible, and punt to the UnixFS spec for how to describe them.

> 

Reviewer 5: couple of other things to consider?

* Directory wrapping at the top level (for just files, kubo has an option to wrap in a directory so you get file metadata)
* Presence and accurate setting of `Tsize` - at one point we were going to deprecate this field for some cases, although I think all our encoders now do it properly, you could just mandate this in the spec though -- all valid profiles must properly encode `Tsize`.

Reviewer: I've added this as a parameter. 

According to the latest version of https://github.com/ipfs/specs/pull/331/files, the calculation is done as follows:

> To compute the Tsize of a child DAG, sum the length of the dag-pb outside message binary length and the blocksizes of all nodes in the child DAG.

If calculated according to this, does it make accurate?

Reviewer 5: sounds about right, I remember there being some nuance in exactly what's included in the size calculation, making it not super stable if you get it slightly wrong (as we did for some variants in go-unixfsnode for a while)

## Detailed design

We introduce a set of named profiles that define a set of parameters for generating UnixFS CIDs. These profiles can be used by implementations to ensure that the same content will yield the same CID across different tools and implementations.

### UnixFS parameters

The profiles define a set of parameters that affect the resulting string encoding of the CID. These parameters are based on the UnixFS specification and are used to generate the CID for a given file tree. The parameters include:

1. CID version, e.g. CIDv0 or CIDv1
1. Multibase encoding for the CID, e.g. base32
1. Hash function used for all nodes in the DAG, e.g. sha2-256
1. UnixFS file chunking algorithm
1. UnixFS file chunk size or target (if required by the chunking algorithm)
1. UnixFS DAG layout (e.g. balanced, trickle etc...)
1. UnixFS DAG width (max number of links per `File` node)
1. `HAMTDirectory` bitwidth, i.e. the number of bits determines the fanout of the `HAMTDirectory` (default bitwidth is 8 == 256 leaves).

> 1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on an estimate of the block size by counting the size of PNNode.Links

Reviewer 7: ```suggestion
1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on an estimate of the block size by counting the size of PNNode.Links. We do not include details about the estimation algorithm as we do not encourage implementations to support it.
```

Reviewer 6: I don't think we should be estimating the block size as it's trivial to calculate it exactly.  Can we not just define this (and punt to the spec for the details) to make it less hand-wavey?

```suggestion
1. `HAMTDirectory` threshold (max `Directory` size before switching to `HAMTDirectory`): based on the final size of the serialized form of the [PBNode protobuf message](https://specs.ipfs.tech/unixfs/#dag-pb-node) that represents the directory.
```

The initial profile in the series, **`unixfs-2025`**, captures the baseline default parameters used by multiple implementations as of November 2025.

| Parameter                     | `unixfs-2025`                                           |
| ----------------------------- | ------------------------------------------------------- |
| CID version                   | CIDv1                                                   |
| Hash function                 | sha2-256                                                |
| Max chunk size                | 1MiB                                                    |
| DAG layout                    | balanced                                                |
| DAG width (children per node) | 1024                                                    |
| `HAMTDirectory` fanout        | 256 blocks                                              |
| `HAMTDirectory` threshold     | 256KiB (estimated by counting the size of PBNode.links) |
| Leaves                        | raw                                                     |

> | Empty directories             | TODO                                                    |

Reviewer 5: ```suggestion
| Empty directories             | TODO                                                    |
| Hidden entities               | TODO                                                    |
```

- 2021-05: [go-unixfs#91](https://github.com/ipfs/go-unixfs/pull/91) introduced `HAMTShardingSize` threshold for automatic sharding based on estimated directory size, using `links-bytes` estimation. This was part of the work tracked in [kubo#8106](https://github.com/ipfs/kubo/issues/8106).

- 2021-11: [go-unixfs#94](https://github.com/ipfs/go-unixfs/pull/94) added size-based unsharding (switching from HAMT back to basic directory), completing bidirectional automatic sharding with `>=` comparison. The [go-unixfs](https://github.com/ipfs/go-unixfs) repository has since been archived; its code now lives in [boxo](https://github.com/ipfs/boxo).

- 2021-12: [go-ipfs v0.11.0](https://github.com/ipfs/kubo/releases/tag/v0.11.0) (now Kubo) shipped with automatic HAMT autosharding, deprecating the global `Experimental.ShardingEnabled` flag.

- 2023-03: boxo created via [Über Migration](https://github.com/ipfs/boxo/commit/038bdd29), inheriting the `>=` comparison behavior from go-unixfs.

- 2026-01: [boxo#1088](https://github.com/ipfs/boxo/pull/1088) fixed threshold comparison from `>=` to `>`, aligning with JS implementation and documentation. Shipped in Kubo 0.40.

> Timeline of JavaScript implementation changes:

Reviewer 6: js-ipfs did link counting, the PR to js-ipfs-unixfs did block-bytes but then [switched](https://github.com/ipfs/js-ipfs-unixfs/pull/171#issuecomment-1424040277) to [link-bytes](https://github.com/ipfs/js-ipfs-unixfs/pull/171/changes/288910ffe735ba5eee9a5007491b9b7ffa4567ab) to be consistent with kubo.
